### PR TITLE
add an ARCHITECTURE.md

### PR DESCRIPTION
**Pull request by lmb**
Describe the high level architecture of the library and explain some
of the choices made so far. The goal is to have a document that only
has to change rarely, so types and files are linked only sparsely.

### CODE REVIEW

**Comment by lmb** [comment]
Inspired by https://github.com/cilium/ebpf/pull/207#issuecomment-777722814 and a conversation with Timo. Consider this a WIP, your feedback would be really valuable to figure out what is missing.

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+Objects
+---
+
+`Program` and `Map` are the result of loading specs into the kernel. Sometimes
+loading a spec will fail because the kernel is too old, or a feature is not
+enabled. There are multiple ways the library deals with that:
+
+* Fallback: older kernels don't allowing naming programs and maps. The library
+  automatically detects support for names, and omits them during load if
+  necessary. This works since names a primarily a debug aid.
```
Comment (MarkPash) [inline ARCHITECTURE.md:53]:
```suggestion
  necessary. This works since names a primarily a debug aid.
```
```suggestion
  necessary. This works since name is primarily a debug aid.
```

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+Architecture of the library
+===
+
+    ELF -> Specifications -> Objects -> Links
+
+ELF
+---
+
+BPF is usually produced by using `clang` to compile a subset of C. `clang` outputs
```
Comment (nathanjsweet) [inline ARCHITECTURE.md:9]: `clang` is a command and "Clang" is a product/tool. I think in most of your cases you should remove the backticks, except in the case of `CollectionSpec` (which should have a godoc link or something).

File: CONTRUBUTING.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 Development is on [GitHub](https://github.com/cilium/ebpf) and contributions in
 the form of pull requests and issues reporting bugs or suggesting new features
-are welcome.
+are welcome. Take a look at [the architecture](ARCHITECTURE.md) of the library.
```
Comment (nathanjsweet) [inline CONTRUBUTING.md:5]: Maybe reword to "Please take a look at the architecture of the library for high-level guidance on the development goals of the library."

**Review by nathanjsweet — CHANGES_REQUESTED** [review]
Some nits. This looks good though. Thanks!

**Comment by lmb** [comment]
Addressed your comments @MarkPash and @nathanjsweet, thanks! Also fleshed out the BTF section.

**Review by nathanjsweet — APPROVED** [review]
